include/deal.II/cgal/triangulation.h

```diff
@@ -0,0 +1,403 @@
+          }
+      }
+    dealii_triangulation.create_triangulation(vertices, cells, subcell_data);
+  }
+
+
+
+  template <typename CGAL_MeshType, int dim, int spacedim>
+  void
+  cgal_surface_to_dealii_triangulation(
+    const CGAL_MeshType &         cgal_mesh,
+    Triangulation<dim, spacedim> &triangulation)
+  {
+    Assert(triangulation.n_cells() == 0,
+           ExcMessage(
+             "Triangulation must be empty upon calling this function."));
+    Assert(dim == 2 && spacedim == 3,
+           ExcMessage("This function works only with dim==2 and spacedim==3."));
+
+
+    auto is_surface_mesh =
+      boost::hana::is_valid([](auto &&obj) -> decltype(obj.faces()) {});
+
+    auto is_polyhedral =
+      boost::hana::is_valid([](auto &&obj) -> decltype(obj.facets_begin()) {});
+
+    // Collect Vertices
+    std::size_t                          n_cgal_vertices;
+    std::vector<dealii::Point<spacedim>> vertices;
+    std::vector<CellData<dim>>           cells;
+    SubCellData                          subcell_data;
+    if constexpr (is_surface_mesh(cgal_mesh))
+      {
+        n_cgal_vertices = cgal_mesh.num_vertices();
+        vertices.reserve(n_cgal_vertices);
+        for (const auto &p : cgal_mesh.points())
+          {
+            vertices.emplace_back(
+              CGALWrappers::cgal_point_to_dealii_point<spacedim>(p));
+          }
+      }
+    else if constexpr (is_polyhedral(cgal_mesh))
+      {
+        n_cgal_vertices = cgal_mesh.size_of_vertices();
+        vertices.reserve(n_cgal_vertices);
+        for (auto it = cgal_mesh.points_begin(); it != cgal_mesh.points_end();
+             ++it)
+          {
+            vertices.emplace_back(
+              CGALWrappers::cgal_point_to_dealii_point<spacedim>(*it));
+          }
+      }
+
+
+    // Different loops for Polyhedron or Surface_mesh types
+    if constexpr (is_surface_mesh(cgal_mesh))
+      {
+        const unsigned int vertices_per_face =
+          CGAL::vertices_around_face(
+            cgal_mesh.halfedge(*(cgal_mesh.faces().begin())), cgal_mesh)
+            .size();
+
+        // Collect CellData
+        for (const auto &face : cgal_mesh.faces())
+          {
+            CellData<dim> c(vertices_per_face);
+            auto          it_vertex = c.vertices.begin();
+            for (const auto v :
+                 CGAL::vertices_around_face(cgal_mesh.halfedge(face),
+                                            cgal_mesh))
+              {
+                *(it_vertex++) = v;
+              }
+
+            if (vertices_per_face == 4)
+              std::swap(c.vertices[3], c.vertices[2]);
+
+            cells.emplace_back(c);
+          }
+        GridTools::delete_unused_vertices(vertices, cells, subcell_data);
+        GridTools::consistently_order_cells(cells);
+        triangulation.create_triangulation(vertices, cells, subcell_data);
+      }
+    else if constexpr (is_polyhedral(cgal_mesh))
+      {
+        const unsigned int vertices_per_face =
```

Comment:
Why is this outside the loop? We support also mixed triangle/quadrilateral meshes. Could you inset this inside the facet loop in line 371 below?